trpcultivate_phenotypes/tests/src/Kernel/Services/Traits/ValidTraitTest.php

```diff
@@ -236,63 +243,435 @@ public function testTraitsServiceDatabaseExpectations() {
+
+      // Set genus to use by the traits service.
+      // This method will return the inserted cvterm ids.
+      $trait_assets = $this->service_traits->insertTrait($ins_trait);
+      // Check trait assets got inserted.
+      $trait_combo = $this->service_traits->getTraitMethodUnitCombo($combo['Trait Name'], $combo['Method Short Name'], $combo['Unit']);
+
+      // Trait.
+      $this->assertEquals($trait_assets['trait'], $trait_combo['trait']->cvterm_id,
+        'Test trait '. $combo['Trait Name'] .' was not inserted.');
+      // Method.
+      $this->assertEquals($trait_assets['method'], $trait_combo['method']->cvterm_id,
+        'Test ' . $combo['Method Short Name'] . ' was not inserted.');
+      // Unit.
+      $this->assertEquals($trait_assets['unit'], $trait_combo['unit']->cvterm_id,
+        'Test unit ' . $combo['Unit'] . ' was not inserted.');
+    }
+
+    // At this point, all traits should be in, test the relationships, connections and all.
+
+    // Test nothing got inserted more than once and re-using a trait asset meant
+    // that it just referenced existing asset and not creating another copy
+    // in the same cv the genus is configured.
+    $reuse_test = [
+      'trait' => 'A Trait',   // Re-used 5x
+      'method' => 'B Method', // Re-used 5x
+      'unit' => 'C Unit'      // Re-used 2x
+    ];
+
+    foreach($reuse_test as $asset => $name) {
+      $exception_message = '';
+
+      try {
+        // Use string trait asset name as parameter.
+        $this->service_traits->getTraitAsset($name, $asset);
+      }
+      catch (Exception $e) {
+        // The getter will throw an exception when it detected multiple copies
+        // of the same name in the same cv the genus is configured.
+        $exception_message = $e->getMessage();
+      }
+
+      $this->assertEmpty($exception_message, 'Test trait asset ' . $asset . ':' . $name . ' was inserted more than once.');
+    }
+
+    // Test that any trait asset can be retrieved either by using
+    // name or id number as parameter using getTraitAsset() method.
+
+    // Test that by using id number and the id provided is an id for a trait asset that belongs
+    // to a cv not the cv set for the genus.
+
+    // term null with id 1 not in the genus.
+    /* Does not seem to catch and would kill the tests.
+    $exception_message = '';
+    try {
+      $this->service_traits->getTraitAsset(1, 'trait');
+    }
+    catch (Exception $e) {
+      $exception_message = $e->getMessage();
+    }
+
+    $this->assertMatchesRegularExpression('/The requested trait asset (trait|method|unit)/', $exception_message,
+      'Trait asset id provided returned an asset not in the cv the genus was configured.');
+    */
+
+    foreach($test_combo as $combo) {
+      foreach($keys as $key => $title) {
+        // Each combo, retrieve assets - trait, method and unit.
+
+        // By string parameter.
+        $asset = $this->service_traits->getTraitAsset($combo[ $title ], $key);
```

Comment:
The getTraitAsset function should have been protected which means we don't test it directly but rather test the results of the other getters.